### PR TITLE
Add datasource ssm_parameters_by_path

### LINKED ISSUE #9429: SSM Parameter store data source which uses GetParametersByPath
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

I think it is good if there is another SSM Parameters Store data source which searches for parameters by calling `GetParametersByPath` API and fetches all the prefixed parameters

### New or Affected Resource(s)

* aws_ssm_parameters_by_path

If other name is better, please change the name of the data source.

### Potential Terraform Configuration

```hcl
data "aws_ssm_parameters_by_path" "foo" {
  path = "/foo"
}
```

### References

* #0000

### LINKED ISSUE #11318: Feature Request: list encrypted or unencrypted aws ssm parameter keys
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Given a path in ssm, e.g. `/env/`

I would like to list all keys under that path, e.g.:

```
/env/FOO
/env/BAR
/env/BAZ
```

of that list of keys, I would like to be able to know which keys are encrypted and which are not

I would like to be able to use a data source to surface that list of keys based on encryption state, so that I may pass the list of unencrypted, and the list of encrypted keys to other resources.

I would use the list of unencrypted keys under `/env/` to then retrieve all the actual values for each key (`/env/FOO`, `/env/BAR`), and store them in an ecs task definition for the environment values.

I would use the list of encrypted keys under `/env/` to provide just the key name, along with the key path, to be used for an ecs task definition for the secrets values.

Basically, this would allow an ssm path consisting of unencrypted and encrypted values, to be turned into lists of environment Key/Value pairs, and secret KeyName/KeyPath pairs.

### New or Affected Resource(s)

* aws_ssm_parameter_keys

### Potential Terraform Configuration

```hcl
/*
given SSM keys:

/env/FOO
/env/BAR
/env/BAZ

where SSM key ending in 'BAZ' is encrypted
*/


# this data source would return only un-encrypted keys
data "aws_ssm_parameter_keys" "env" {
  prefix = "/env/"
  encrypted = false
}

/*
this would result in:

data.aws_ssm_parameter_keys.env = list(string)

or

data.aws_ssm_parameter_keys.env = [
  "/env/FOO",
  "/env/BAR",
]
*/

# this data source would return only encrypted keys
data "aws_ssm_parameter_keys" "secrets" {
  prefix = "/env/"
  encrypted = true
}

/*
this would result in:

data.aws_ssm_parameter_keys.secrets = list(string)

or

data.aws_ssm_parameter_keys.secrets = [
  "/env/BAZ",
]
*/

# get the values for all the unencrypted keys
data "aws_ssm_parameter" "env_values" {
  for_each = data.aws_ssm_parameter_keys.env

  name = each.value
}

locals {
  /*
  convert unencrypted env keys into env var -> values:

  ssm_env = {
    "FOO": "value_from_ssm1",
    "BAR": "value_from_ssm2",
  }
  */

  ssm_env = {for value in data.aws_ssm_parameter.env_values: split("/", value.name)[-1] => value.value}

  /*
  convert encrypted secret keys into env var -> ssm key:

  ssm_env = {
    "BAZ": "/env/BAZ",
  }
  */
  ssm_secrets = {for k, v in data.aws_ssm_parameter_keys.secrets : split("/", k)[-1] => k}
}

resource "aws_ecs_task_definition" "web" {
  family = "foo"

  container_definitions = <<-EOL
  [
    {
      # use the environment values and secret key paths from above
      # to inform the container definition
      "name": "foo",
      "environment" : ${jsonencode(local.ssm_env)},
      "secrets": ${jsonencode([for k, v in local.ssm_secrets : {name=k,valueFrom=v}])},
    }
  ]
  EOL
}
```

### References

N\A

### PR DESCRIPTION
**Pull request by katainaka0503**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Closes #9429.
Closes #11318.

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
New Data Source: aws_ssm_parameters_by_path
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSSsmParametersByPathDataSource_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -parallel 20 -run=TestAccAWSSsmParametersByPathDataSource_basic -timeout 120m
?       github.com/terraform-providers/terraform-provider-aws   [no test files]
=== RUN   TestAccAWSSsmParametersByPathDataSource_basic
=== PAUSE TestAccAWSSsmParametersByPathDataSource_basic
=== CONT  TestAccAWSSsmParametersByPathDataSource_basic
--- PASS: TestAccAWSSsmParametersByPathDataSource_basic (32.50s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       32.551s
```

### CODE REVIEW

File: aws/provider.go
Comment (katainaka0503) [inline aws/provider.go]: I could not determine which naming is better `aws_ssm_parameters_by_path` or simply `aws_ssm_parameters`

**Comment by cmeisinger** [comment]
Any updates here?

**Comment by Dgadavin** [comment]
Hey guys. This is very useful things. I'm very appreciate any movement in this PR.

**Comment by DmitriyLyalyuev** [comment]
Guys, we need this. Please merge :)

File: aws/provider.go
Comment (robertcnix) [reply aws/provider.go:264]: IMO `aws_ssm_parameters` is cleaner but I'd prefer aligning with the aws-cli so `aws_ssm_parameters_by_path` is the wiser choice.

**Comment by katainaka0503** [comment]
I've resolved a conflict.

**Comment by qliam** [comment]
do you know when should be release please ? It would be very useful features :)

**Comment by teamterraform** [comment]
## Notification of Recent and Upcoming Changes to Contributions

Thank you for this contribution! There have been a few recent development changes that affect this pull request. We apologize for the inconvenience, especially if there have been long review delays up until now. Please note that this is automated message from an unmonitored account. See the [FAQ](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/FAQ.md) for additional information on the maintainer team and review prioritization.

If you are unable to complete these updates, please leave a comment for the community and maintainers so someone can potentially continue the work. The maintainers will encourage other contributors to use the existing contribution as the base for additional changes as appropriate. Otherwise, contributions that do not receive updated code or comments from the original contributor may be closed in the future so the maintainers can focus on active items.

For the most up to date information about Terraform AWS Provider development, see the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md). Additional technical debt changes can be tracked with the [`technical-debt` label on issues](https://github.com/terraform-providers/terraform-provider-aws/issues?q=is%3Aopen+is%3Aissue+label%3Atechnical-debt).

As part of updating a pull request with these changes, the most current unit testing and linting will run. These may report issues that were not previously reported.

### Terraform 0.12 Syntax

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/8950
Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14417

Version 3 and later of the Terraform AWS Provider, which all existing contributions would potentially be added, only supports Terraform 0.12 and later. Certain syntax elements of Terraform 0.11 and earlier show deprecation warnings during runs with Terraform 0.12. Documentation and test configurations, such as those including deprecated string interpolations (`some_attribute = "${aws_service_thing.example.id}"`) should be updated to the newer syntax (`some_attribute = aws_service_thing.example.id`). Contribution testing will automatically fail on older syntax in the near future. Please see the referenced issues for additional information.

### Action Required: Terraform Plugin SDK Version 2

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14551

The Terraform AWS Provider has been upgraded to the latest version of the Terraform Plugin SDK. Generally, most changes to contributions should only involve updating Go import paths in source code files. Please see the referenced issue for additional information.

### Action Required: Removal of website/aws.erb File

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14712

Any changes to the `website/aws.erb` file are no longer necessary and should be removed from this contribution to prevent merge issues in the near future when the file is removed from the repository. Please see the referenced issue for additional information.

### Upcoming Change of Git Branch Naming

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14292

Development environments will need their upstream Git branch updated from `master` to `main` in the near future. Please see the referenced issue for additional information and scheduling.

### Upcoming Change of GitHub Organization

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14715

This repository will be migrating from https://github.com/terraform-providers/terraform-provider-aws to https://github.com/hashicorp/terraform-provider-aws. No practitioner or developer action is anticipated and most GitHub functionality will automatically redirect to the new location. Go import paths including `terraform-providers` can remain for now. Please see the referenced issue for additional information and scheduling.

**Comment by tsalinguepumpkin** [comment]
@katainaka0503 do you know if this gonna be merged ?

**Comment by katainaka0503** [comment]
@tsalinguepumpkin I don't know how thiis repo's mainteners wish to deal with this PR at all.

**Comment by melvinkcx** [comment]
We need this feature too! The workaround I have is utterly ugly RN

**Comment by maulik-vimond** [comment]
Is there any chance that this will be merged any time soon? This is a very useful feature.

**Comment by itays333** [comment]
Was surprised to find out this feature doesn't exist yet, I need this as well.

**Comment by theuves** [comment]
This is very useful! Merge it, please!!!

**Comment by tsalinguepumpkin** [comment]
any updates on the subject ?

**Comment by fatbasstard** [comment]
Just found this PR since I also "need" this datasource. Any chance this can/will get any support?

@katainaka0503 You're still monitoring this?

**Comment by katainaka0503** [comment]
Yes. If maintainors are still monitoring this PR, I'm ready to rewrite this PR as it fits current code base.

**Comment by fatbasstard** [comment]
Going to tag some of the recent contibuters: @anGie44 @ewbankkit @bflad 

This is feature people want (including myself 😉 ) and @katainaka0503 is willing to finish up the PR. Can this PR get some ❤️ ?

**Comment by yuklia** [comment]
hi! any news? this feature a really great and we all waiting for the release 👍

File: aws/provider.go
Comment (dominik-ba) [reply aws/provider.go:315]: differentiating this feature and the single parameter only with one trailing "s" sounds like problematic user/developer experience :)

**Comment by jbcom** [comment]
Is there anything I could do to help with this? Right now the approach we're using is hacky as heck involving awscli and I would massively love to not be doing that.

**Comment by nickperkins** [comment]
@hc-github-team-terraform-aws anything needed for this to be reviewed?

**Comment by nickperkins** [comment]
> Yes. If maintainors are still monitoring this PR, I'm ready to rewrite this PR as it fits current code base.

It might be worth closing this and creating a new PR with the updates rom https://github.com/hashicorp/terraform-provider-aws/pull/9615#issuecomment-675685375.

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

##### Commercial

```console
% make testacc TESTARGS='-run=TestAccAWSSsmParametersByPathDataSource_'        
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSsmParametersByPathDataSource_ -timeout 180m
=== RUN   TestAccAWSSsmParametersByPathDataSource_basic
=== PAUSE TestAccAWSSsmParametersByPathDataSource_basic
=== CONT  TestAccAWSSsmParametersByPathDataSource_basic
--- PASS: TestAccAWSSsmParametersByPathDataSource_basic (17.70s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	20.858s
```

##### GovCloud

```console
% make testacc TESTARGS='-run=TestAccAWSSsmParametersByPathDataSource_' 
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSsmParametersByPathDataSource_ -timeout 180m
=== RUN   TestAccAWSSsmParametersByPathDataSource_basic
=== PAUSE TestAccAWSSsmParametersByPathDataSource_basic
=== CONT  TestAccAWSSsmParametersByPathDataSource_basic
--- PASS: TestAccAWSSsmParametersByPathDataSource_basic (20.37s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	25.193s
```

**Comment by ewbankkit** [comment]
@katainaka0503 Thanks for the contribution :tada: :clap:.

**Pull request merged**
